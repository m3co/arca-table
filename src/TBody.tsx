
import * as React from 'react';

import { TRow } from './TRow';

import { Field, Row, Info } from 'arca-redux';

import { SearchSocket } from './Search-socket';

interface Props {
  newRow?: Row;
  onDeleteNewRow?: (Row: Row) => void;
  Info: Info;
  Rows: Row[];
  onUpdate?: (Row: Row, column?: keyof Row, Field?: Field) => void;
  onInsert?: (Row: Row, column?: keyof Row, Field?: Field) => void;
  onDelete?: (Row: Row) => void;
}

export class TBody
  extends React.Component<Props>
{
  private socket?: SearchSocket;
  public constructor(props: Props) {
    super(props);

    let shouldOpenSocket = !!props.Info.Fields
      .find((field): boolean => !!field.Combobox);
    if (shouldOpenSocket) {
      this.socket = new SearchSocket();
    }
  }

  public render(): JSX.Element {
    const { Info, Rows, newRow,
      onUpdate, onInsert, onDelete, onDeleteNewRow } = this.props;
    return (
      <tbody>
        {
          newRow &&
          <TRow
            key={'newRow'}
            Info={Info}
            Row={newRow}
            onEdit={onInsert}
            onRemove={onDeleteNewRow}
            dirty={true}
            socket={this.socket} />
        }
        {
          Rows.map((Row, key): JSX.Element => {
            return (<TRow
              key={key}
              Info={Info}
              Row={Row}
              onEdit={onUpdate}
              onRemove={onDelete}
              socket={this.socket} />);
          })
        }
      </tbody>
    );
  }
}
