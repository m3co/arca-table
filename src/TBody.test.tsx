
import * as React from 'react';
import { mount } from 'enzyme';

import Socket from 'socket.io-client';
import { ARCASocket, reducer, FACADReports } from 'arca-redux';
import { createStore } from 'redux';

import { TBody } from './TBody';

const URL = 'http://localhost:8086';
const Source = 'FACAD-Reports';

test('TBody renders FACAD-Reports', (done): void => {
  const io = Socket(URL);
  const store = createStore(reducer);
  const socket = new ARCASocket(store, io);

  socket.GetInfo(Source);
  socket.Select(Source);
  store.subscribe((): void => {
    const state = store.getState();
    const { Rows, Info } = state.Source[Source];
    if (state.Connected &&
      Info &&
      Rows.length > 0) {

      const el = (
        <table>
          <TBody
            Rows={Rows}
            Info={Info} />
        </table>
      );
      const wrapper = mount(el);
      expect(wrapper).toMatchSnapshot();
      done();
    }
  });
});

test('TBody adds a new row FACAD-Reports', (done): void => {
  const io = Socket(URL);
  const store = createStore(reducer);
  const socket = new ARCASocket(store, io);

  const newRow: FACADReports["Row"] = {
    ID: 0,
    BuiltInCategory: 'INVALID',
    ReportType: 'Schedule',
    Name: 'A name',
    PathName: 'Path Name',
    Field1: '',
    Field2: '',
    Field3: '',
  };

  socket.GetInfo(Source);
  socket.Select(Source);
  store.subscribe((): void => {
    const state = store.getState();
    const { Rows, Info } = state.Source[Source];
    if (state.Connected &&
      Info &&
      Rows.length > 0) {

      const el = (
        <table>
          <TBody
            Rows={Rows}
            newRow={newRow}
            Info={Info} />
        </table>
      );
      const wrapper = mount(el);
      expect(wrapper).toMatchSnapshot();
      done();
    }
  });
});
