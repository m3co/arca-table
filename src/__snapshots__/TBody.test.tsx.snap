// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TBody adds a new row FACAD-Reports 1`] = `
<table>
  <TBody
    Info={
      Object {
        "Actions": Object {
          "Delete": true,
          "Insert": true,
          "Update": true,
        },
        "Fields": Array [
          Object {
            "Combobox": null,
            "Editable": false,
            "Name": "ID",
            "Primary": true,
            "Required": false,
            "Select": null,
            "Type": "Number",
          },
          Object {
            "Combobox": null,
            "Editable": true,
            "Name": "Name",
            "Primary": false,
            "Required": true,
            "Select": null,
            "Type": "String",
          },
          Object {
            "Combobox": null,
            "Editable": true,
            "Name": "BuiltInCategory",
            "Primary": false,
            "Required": true,
            "Select": null,
            "Type": "String",
          },
          Object {
            "Combobox": null,
            "Editable": true,
            "Name": "PathName",
            "Primary": false,
            "Required": true,
            "Select": null,
            "Type": "String",
          },
        ],
      }
    }
    Rows={
      Array [
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 19,
          "Name": "1ssssxdd",
          "PathName": "12121sdgcw",
        },
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 22,
          "Name": "csdcs",
          "PathName": "sdsd",
        },
      ]
    }
    newRow={
      Object {
        "BuiltInCategory": "INVALID",
        "Field1": "",
        "Field2": "",
        "Field3": "",
        "ID": 0,
        "Name": "A name",
        "PathName": "Path Name",
        "ReportType": "Schedule",
      }
    }
  >
    <tbody>
      <TRow
        Info={
          Object {
            "Actions": Object {
              "Delete": true,
              "Insert": true,
              "Update": true,
            },
            "Fields": Array [
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
            ],
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 0,
            "Name": "A name",
            "PathName": "Path Name",
            "ReportType": "Schedule",
          }
        }
        dirty={true}
        key="newRow"
      >
        <tr
          className="dirty"
        >
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "Field1": "",
                "Field2": "",
                "Field3": "",
                "ID": 0,
                "Name": "A name",
                "PathName": "Path Name",
                "ReportType": "Schedule",
              }
            }
            key="0"
          >
            <td
              className="primary"
              onClick={[Function]}
            >
              0
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "Field1": "",
                "Field2": "",
                "Field3": "",
                "ID": 0,
                "Name": "A name",
                "PathName": "Path Name",
                "ReportType": "Schedule",
              }
            }
            key="1"
          >
            <td
              onClick={[Function]}
            >
              A name
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "Field1": "",
                "Field2": "",
                "Field3": "",
                "ID": 0,
                "Name": "A name",
                "PathName": "Path Name",
                "ReportType": "Schedule",
              }
            }
            key="2"
          >
            <td
              onClick={[Function]}
            >
              INVALID
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "Field1": "",
                "Field2": "",
                "Field3": "",
                "ID": 0,
                "Name": "A name",
                "PathName": "Path Name",
                "ReportType": "Schedule",
              }
            }
            key="3"
          >
            <td
              onClick={[Function]}
            >
              Path Name
            </td>
          </TCell>
          <td>
            <button
              onClick={[Function]}
            >
              -
            </button>
          </td>
        </tr>
      </TRow>
      <TRow
        Info={
          Object {
            "Actions": Object {
              "Delete": true,
              "Insert": true,
              "Update": true,
            },
            "Fields": Array [
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
            ],
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "ID": 19,
            "Name": "1ssssxdd",
            "PathName": "12121sdgcw",
          }
        }
        key="0"
      >
        <tr>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="0"
          >
            <td
              className="primary"
              onClick={[Function]}
            >
              19
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="1"
          >
            <td
              onClick={[Function]}
            >
              1ssssxdd
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="2"
          >
            <td
              onClick={[Function]}
            >
              INVALID
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="3"
          >
            <td
              onClick={[Function]}
            >
              12121sdgcw
            </td>
          </TCell>
          <td>
            <button
              onClick={[Function]}
            >
              -
            </button>
          </td>
        </tr>
      </TRow>
      <TRow
        Info={
          Object {
            "Actions": Object {
              "Delete": true,
              "Insert": true,
              "Update": true,
            },
            "Fields": Array [
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
            ],
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "ID": 22,
            "Name": "csdcs",
            "PathName": "sdsd",
          }
        }
        key="1"
      >
        <tr>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="0"
          >
            <td
              className="primary"
              onClick={[Function]}
            >
              22
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="1"
          >
            <td
              onClick={[Function]}
            >
              csdcs
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="2"
          >
            <td
              onClick={[Function]}
            >
              INVALID
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="3"
          >
            <td
              onClick={[Function]}
            >
              sdsd
            </td>
          </TCell>
          <td>
            <button
              onClick={[Function]}
            >
              -
            </button>
          </td>
        </tr>
      </TRow>
    </tbody>
  </TBody>
</table>
`;

exports[`TBody renders FACAD-Reports 1`] = `
<table>
  <TBody
    Info={
      Object {
        "Actions": Object {
          "Delete": true,
          "Insert": true,
          "Update": true,
        },
        "Fields": Array [
          Object {
            "Combobox": null,
            "Editable": false,
            "Name": "ID",
            "Primary": true,
            "Required": false,
            "Select": null,
            "Type": "Number",
          },
          Object {
            "Combobox": null,
            "Editable": true,
            "Name": "Name",
            "Primary": false,
            "Required": true,
            "Select": null,
            "Type": "String",
          },
          Object {
            "Combobox": null,
            "Editable": true,
            "Name": "BuiltInCategory",
            "Primary": false,
            "Required": true,
            "Select": null,
            "Type": "String",
          },
          Object {
            "Combobox": null,
            "Editable": true,
            "Name": "PathName",
            "Primary": false,
            "Required": true,
            "Select": null,
            "Type": "String",
          },
        ],
      }
    }
    Rows={
      Array [
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 19,
          "Name": "1ssssxdd",
          "PathName": "12121sdgcw",
        },
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 22,
          "Name": "csdcs",
          "PathName": "sdsd",
        },
      ]
    }
  >
    <tbody>
      <TRow
        Info={
          Object {
            "Actions": Object {
              "Delete": true,
              "Insert": true,
              "Update": true,
            },
            "Fields": Array [
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
            ],
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "ID": 19,
            "Name": "1ssssxdd",
            "PathName": "12121sdgcw",
          }
        }
        key="0"
      >
        <tr>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="0"
          >
            <td
              className="primary"
              onClick={[Function]}
            >
              19
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="1"
          >
            <td
              onClick={[Function]}
            >
              1ssssxdd
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="2"
          >
            <td
              onClick={[Function]}
            >
              INVALID
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 19,
                "Name": "1ssssxdd",
                "PathName": "12121sdgcw",
              }
            }
            key="3"
          >
            <td
              onClick={[Function]}
            >
              12121sdgcw
            </td>
          </TCell>
          <td>
            <button
              onClick={[Function]}
            >
              -
            </button>
          </td>
        </tr>
      </TRow>
      <TRow
        Info={
          Object {
            "Actions": Object {
              "Delete": true,
              "Insert": true,
              "Update": true,
            },
            "Fields": Array [
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              },
            ],
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "ID": 22,
            "Name": "csdcs",
            "PathName": "sdsd",
          }
        }
        key="1"
      >
        <tr>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": false,
                "Name": "ID",
                "Primary": true,
                "Required": false,
                "Select": null,
                "Type": "Number",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="0"
          >
            <td
              className="primary"
              onClick={[Function]}
            >
              22
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "Name",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="1"
          >
            <td
              onClick={[Function]}
            >
              csdcs
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "BuiltInCategory",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="2"
          >
            <td
              onClick={[Function]}
            >
              INVALID
            </td>
          </TCell>
          <TCell
            Field={
              Object {
                "Combobox": null,
                "Editable": true,
                "Name": "PathName",
                "Primary": false,
                "Required": true,
                "Select": null,
                "Type": "String",
              }
            }
            Row={
              Object {
                "BuiltInCategory": "INVALID",
                "ID": 22,
                "Name": "csdcs",
                "PathName": "sdsd",
              }
            }
            key="3"
          >
            <td
              onClick={[Function]}
            >
              sdsd
            </td>
          </TCell>
          <td>
            <button
              onClick={[Function]}
            >
              -
            </button>
          </td>
        </tr>
      </TRow>
    </tbody>
  </TBody>
</table>
`;
